language/reserved_keywords.rb

```diff
@@ -0,0 +1,159 @@
+require_relative '../spec_helper'
+
+describe "Ruby's reserved keywords" do
+  # Copied from Prism::Translation::Ripper
+  keywords = %w[
+    alias
+    and
+    begin
+    BEGIN
+    break
+    case
+    class
+    def
+    defined?
+    do
+    else
+    elsif
+    end
+    END
+    ensure
+    false
+    for
+    if
+    in
+    module
+    next
+    nil
+    not
+    or
+    redo
+    rescue
+    retry
+    return
+    self
+    super
+    then
+    true
+    undef
+    unless
+    until
+    when
+    while
+    yield
+    __ENCODING__
+    __FILE__
+    __LINE__
+  ]
+
+  keywords.each do |kw|
+    describe "keyword '#{kw}'" do
+      it "can't be used as local variable name" do
+        -> { eval(<<~RUBY) }.should raise_error(SyntaxError)
+            #{kw} = "a local variable named '#{kw}'"
+        RUBY
+      end
+
+      invalid_ivar_names = ["defined?"]
+
+      if invalid_ivar_names.include?(kw)
+        it "can't be used as an instance variable name" do
+          -> { eval(<<~RUBY) }.should raise_error(SyntaxError)
+            @#{kw} = "an instance variable named '#{kw}'"
+          RUBY
+        end
+      else
+        it "can be used as an instance variable name" do
+          result = eval <<~RUBY
+            @#{kw} = "an instance variable named '#{kw}'"
+            @#{kw}
+          RUBY
+
+          result.should == "an instance variable named '#{kw}'"
+        end
+      end
+
+      invalid_class_var_names = ["defined?"]
+
+      if invalid_class_var_names.include?(kw)
+        it "can't be used as a class variable name" do
+          -> { eval(<<~RUBY) }.should raise_error(SyntaxError)
+            class C
+              @@#{kw} = "a class variable named '#{kw}'"
+            end
+          RUBY
+        end
+      else
+        it "can be used as a class variable name" do
+          result = eval <<~RUBY
+            class C
+              @@#{kw} = "a class variable named '#{kw}'"
+              @@#{kw}
+            end
+          RUBY
+
+          result.should == "a class variable named '#{kw}'"
+        end
+      end
+
+      invalid_global_var_names = ["defined?"]
```

Comment:
Since it's the same invalid for ivar, cvar and gvar I would like to make that clear in the structure of the spec.
Like:
```ruby
it "can't be used as local variable name" do

if kw == 'defined'
  it "can't be used as a instance variable name" do
  it "can't be used as a class variable name" do
  it "can't be used as a global variable name" do
else
  it "can be used as a instance variable name" do
  it "can be used as a class variable name" do
  it "can be used as a global variable name" do
end
```